### PR TITLE
Implement `$type` aggregation operator

### LINKED ISSUE #2678: Implement `$type` aggregation operator
### What should be done?

https://www.mongodb.com/docs/manual/reference/operator/aggregation/type/

Currently we have a stub for `$type` operator so it just needs an implementation.
The `commonparam` package offers the `AliasFromType` function which can be used for the operator either.

#2635 might be helpful, there's a basic implementation there and couple of test cases.

As `$type` differentiate between unset and Null, we should probably modify `expression.Evaluate` to return nil and not `Null` for non-existent paths. Some changes in code will be needed to handle function calls in a different way.

As we currently don't support standard operators anywhere, let's add basic operators support in the `$project` stage. Other stages will be covered in #2679 after implementing `$type`.

### DoD
- `$type` operator is supported
- recursive operator calls in `$type` are supported (for example `{$type:{$type:42}}`)
- dot notation is supported and tested properly (arrays, and array indexes should be also tested)
- compat tests implemented.

### PR DESCRIPTION
**Pull request by noisersup**
# Description

This PR is a first part of this issue. More advanced error handling for recursive calls will be introduce in separate PR to not overcomplicate the review process.

Closes #2678.

## Readiness checklist

- [ ] I added/updated unit tests.
- [x] I added/updated integration/compatibility tests.
- [x] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/handlers/common/aggregations/operators/operators.go [hunk 1]
```diff
@@ -59,12 +69,38 @@ type Operator interface {
...
+
+		if strings.HasPrefix(k, "$") {
+			operatorExists = true
+			break
+		}
+	}
+
+	switch {
+	case !operatorExists:
+		return nil, ErrNoOperator
```
[the remark this answers is not in the document]
Comment (noisersup) [reply internal/handlers/common/aggregations/operators/operators.go:100]: My main motivation was the amount of repetitive code needed to parse the operator.

File: internal/handlers/common/aggregations/expression.go [hunk 2]
```diff
@@ -160,16 +160,15 @@ func (e *Expression) Evaluate(doc *types.Document) any {
 	if len(vals) == 0 {
 		if isPrefixArray {
 			// when the prefix is array, return empty array.
-			return must.NotFail(types.NewArray())
+			return must.NotFail(types.NewArray()), nil
 		}
 
-		// $group stage groups non-existent paths with `Null`
-		return types.Null
+		return nil, fmt.Errorf("no document found under %s path", path)
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/expression.go:166]: Why not create an `ExpressionError`?

File: internal/handlers/common/aggregations/expression.go [hunk 3]
```diff
@@ -132,18 +133,17 @@ func NewExpression(expression string) (*Expression, error) {
...
+// It returns error if the path does not exists.
+func (e *Expression) Evaluate(doc *types.Document) (any, error) {
 	path := e.path
 
 	if path.Len() == 1 {
 		val, err := doc.Get(path.String())
 		if err != nil {
-			// $group stage groups non-existent paths with `Null`
-			return types.Null
+			return nil, err
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/expression.go:143]: We can create `ExpressionError` right?

File: internal/handlers/common/aggregations/operators/operators.go [hunk 4]
```diff
@@ -59,12 +66,38 @@ type Operator interface {
 // The document should look like: `{<$operator>: <operator-value>}`.
 func NewOperator(doc any) (Operator, error) {
 	operatorDoc, ok := doc.(*types.Document)
-
-	switch {
-	case !ok:
+	if !ok {
 		return nil, ErrWrongType
-	case operatorDoc.Len() == 0:
+	}
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/operators/operators.go:71]: Could you add todo to cover the test please? `// TODO: https://github.com/FerretDB/FerretDB/pull/2789`

File: internal/handlers/common/aggregations/operators/operators.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/common/aggregations/operators/operators.go:100]: Thanks, let's keep it this way for now :hugs:

File: integration/aggregate_documents_compat_test.go [hunk 5]
```diff
@@ -1597,12 +1597,98 @@ func TestAggregateCompatProject(t *testing.T) {
...
+			pipeline: bson.A{
+				bson.D{{"$sort", bson.D{{"_id", -1}}}},
+				bson.D{{"$project", bson.D{{"type", bson.D{{"$type", primitive.NewObjectID()}}}}}},
+			},
+		},
+		"TypeBool": {
+			pipeline: bson.A{
+				bson.D{{"$sort", bson.D{{"_id", -1}}}},
+				bson.D{{"$project", bson.D{{"type", bson.D{{"$type", true}}}}}},
+			},
```
Comment (chilagrow) [inline integration/aggregate_documents_compat_test.go:1691]: Could you also test float64, types.Binary, time.Time, types.NullType, types.Regex, types.Timestamp please :pray:

File: internal/handlers/common/aggregations/operators/type.go [hunk 6]
```diff
@@ -16,37 +16,100 @@
...
+			}
+
+			return commonparams.AliasFromType(param), nil
+
+		default:
+			panic(fmt.Sprint("wrong type of value: ", typeParam))
+		}
+	}
+
+	return nil, nil
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/operators/type.go:114]: Is this reachable?

Comment (noisersup) [reply internal/handlers/common/aggregations/operators/type.go]: Fixed

File: internal/handlers/common/aggregations/expression.go (hunk 3 above)
Comment (chilagrow) [reply internal/handlers/common/aggregations/expression.go:143]: I assume this one is the same as https://github.com/FerretDB/FerretDB/pull/2789/files#r1227739197

File: internal/handlers/common/aggregations/expression.go (hunk 2 above)
Comment (chilagrow) [reply internal/handlers/common/aggregations/expression.go:166]: Ok

**Review by chilagrow — APPROVED** [review]

**Review by rumyantseva — APPROVED** [review]
LGTM!

**Pull request merged**